sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -299,21 +299,49 @@ private static Method findAnnotatedMethod(
     }
 
     // We need to be able to call it. We require it is public.
-    checkArgument(
+    errors.checkArgument(
         (first.getModifiers() & Modifier.PUBLIC) != 0, "%s must be public", format(first));
 
     // And make sure its not static.
-    checkArgument(
+    errors.checkArgument(
         (first.getModifiers() & Modifier.STATIC) == 0, "%s must not be static", format(first));
 
     return first;
   }
 
   private static String format(Method m) {
-    return ReflectHelpers.CLASS_AND_METHOD_FORMATTER.apply(m);
+    return (m == null) ? "null" : ReflectHelpers.CLASS_AND_METHOD_FORMATTER.apply(m);
   }
 
   private static String formatType(TypeToken<?> t) {
-    return ReflectHelpers.TYPE_SIMPLE_DESCRIPTION.apply(t.getType());
+    return (t == null) ? "null" : ReflectHelpers.TYPE_SIMPLE_DESCRIPTION.apply(t.getType());
+  }
+
+  static class ErrorReporter {
+    private final String label;
+
+    ErrorReporter(@Nullable ErrorReporter root, String label) {
+      this.label = (root == null) ? label : String.format("%s, %s", root.label, label);
+    }
+
+    ErrorReporter forMethod(String annotation, Method method) {
+      return new ErrorReporter(this, String.format("%s method %s", annotation, format(method)));
```

Comment:
Could take the annotation as a class, and do `String.format("@%s(%s)", annotation.getSimpleName(), format(method))` here, to make the `forMethod` less likely to encounter a type.